### PR TITLE
Send new install ID and install type headers with telemetry messages

### PR DESCRIPTION
**Pull request by yshapiro-57**
### What does this PR do?

This adds two new headers to telemetry messages, one that uniquely identifies the agent installation and another one that shows how the customer installed the agent on their host (manual, installation script, single-step APM etc.)

### Motivation

Jira: [Send additional tags from the Datadog agent telemetry events](https://datadoghq.atlassian.net/browse/AIT-8905)
RFC: [APM Onboarding time-to-value telemetry](https://docs.google.com/document/d/14vsrCbnAKnXmJAkacX9I6jKPGKmxsq0PKUb3dfiZpWE/edit#heading=h.8d3o7vtyu1y1)

### Describe how to test/QA your changes

Make sure all Circle CI tests are green, meaning that the new unit tests all pass.

For an end to end test, make sure your local agent config in `/opt/datadog-agent/etc/datadog.yaml` has `dd_url: https://ddstaging.datadoghq.com` and a valid API key from staging. Build and run the trace agent locally on your MacBook with these commands:

```
cd ~/dd/datadog-agent/
go build ./cmd/trace-agent
./trace-agent -c /opt/datadog-agent/etc/datadog.yaml
```

Verify that this generates a `/opt/datadog-agent/etc/install.json` with contents like this:
```
{"install_id":"00000014-7f4a-21ee-9b01-4e8b84448778","install_type":"manual","install_time":1699566719}
```

Run an instrumented service locally on your MacBook. Verify that the first trace that gets sent through the trace agent is tagged with the `_dd.install.id`, `_dd.install.type` and `_dd.install.time` tags that match what's in the install.json.

Repeat this test, but now write your own values for install_id, install_type and install_time in the `install.json` file before you fire up the agent. (You can use any UUID, any string, and UNIX timestamp.) Make sure the agent does not clobber the values that you wrote to `install.json`. Make sure the values you wrote to `install.json` are still tagged onto the first trace that's sent by the trace agent.

After you have run this test, post in our team's Slack channel (#instrumentation-telemetry) to have someone from the Telemetry team verify for you that your test resulted in an `agent.first_trace.sent` telemetry event being received and processed by the Datadog backend in our staging environment.

https://datadoghq.atlassian.net/browse/AIT-8951

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/trace/config/setup.go
Comment (knusbaum) [inline comp/trace/config/setup.go:356]: This can end up with us having a partial signature, I believe. e.g. `InstallType` but no time or id.

File: pkg/trace/telemetry/collector.go
Comment (knusbaum) [inline pkg/trace/telemetry/collector.go:147]: Looks like this will return an error even if some endpoint succeeded. Also it only returns the last error.

I see below we're checking if we've sent before. If there's a bad endpoint in the config, I think this will result in constant sending of first trace events.

**Review by knusbaum — CHANGES_REQUESTED** [review]
Couple minor things.

File: comp/trace/config/setup.go
Comment (paullegranddc) [reply comp/trace/config/setup.go:356]: True, I don't think it's that's a big issue because we'll still make some sense of the telemetry data with only one or two of them.
Also the goal is that these configs won't be set by users but either injected into the env/config by the operator/helm chat or set by the installer

File: pkg/trace/telemetry/collector.go
Comment (paullegranddc) [reply pkg/trace/telemetry/collector.go:147]: I just don't want us to drop the event if one of the endpoints has temporary issues

We could hope that this is not too common, or add limited retries

Comment (knusbaum) [reply pkg/trace/telemetry/collector.go:147]: Is it a problem that it's going to be submitted multiple times to the same endpoint?
If one endpoint fails, the successful endpoints will get continuous events until it stops failing. Not sure of your use case for the first trace event, but I would think it would make it more or less useless.

It's not that uncommon to have misconfigurations, or services being down. I think we should account for this.

At the very least, we should limit this in some way. We don't want a bad config or down service to trigger a telemetry submission on every trace write.

Comment (yshapiro-57) [reply pkg/trace/telemetry/collector.go:154]: All of our endpoints are idempotent on the backend: one and the same event is safe to send any number of times, the backend will correctly deduplicate them based on a hash of their contents.

You do have a good point that if one of the endpoints is down, the other one will get spammed with continual "first trace sent" errors. I've made a change to try sending the event no more than 5 times. If we fail, we bump a counter and once it reaches a certain count, we stop retrying.

Comment (knusbaum) [reply pkg/trace/telemetry/collector.go:154]: > All of our endpoints are idempotent on the backend: one and the same event is safe to send any number of times, the backend will correctly deduplicate them based on a hash of their contents.

Ok, good to know. It looks like time isn't a parameter, so it will be the same payload every time, even if it's triggered later or by a different payload?

> You do have a good point that if one of the endpoints is down, the other one will get spammed with continual "first trace sent" errors. I've made a change to try sending the event no more than 5 times. If we fail, we bump a counter and once it reaches a certain count, we stop retrying.

This looks fine.

Comment (yshapiro-57) [reply pkg/trace/telemetry/collector.go:154]: Correct, this is just the time the agent was originally installed, not the time the trace is being sent.

**Review by knusbaum — APPROVED** [review]

**Review by paullegranddc — APPROVED** [review]

**Review by billyac — APPROVED** [review]

**Review by darccio — APPROVED** [review]
Approved with some pending small changes.

File: comp/trace/config/install_signature.go
Comment (darccio) [inline comp/trace/config/install_signature.go:39]: Move on top of the previous code block.

File: pkg/trace/agent/agent.go
Comment (darccio) [inline pkg/trace/agent/agent.go:357]: Move this below, after the block that checks if the trace is dropped or not.

File: pkg/trace/agent/agent.go
Comment (knusbaum) [inline pkg/trace/agent/agent.go:356]:
```suggestion
			// Now that we know this trace has been sampled,
```

**Review by knusbaum — APPROVED** [review]
Looks good.

One documentation nitpick.

**Review by billyac — APPROVED** [review]

**Review by DylanLovesCoffee — APPROVED** [review]
✅

**Comment by yshapiro-57** [comment]
Per my conversation with @DylanLovesCoffee and @alexgallotta the CI failures in Serverless Integration Tests / appsec are flaky and OK to ignore, as these tests have been added only recently. I am waiting on the long-running tests to complete before merging this to main.

**Pull request merged**
